### DISCUSSION TITLE
Which is the recommended way? ‘shallow’ or ‘useShallow'

### DISCUSSION THREAD

**Discussion by Chuck-Ray** [pmndrs/zustand#2203] [Q&A] [answered]
Hello.

When I was checking the community comments, I saw that the recommended writing style was the following:

```javascript

createWithEqualityFn((state)=>{}, shallow);

```

However, I am mentioned on the home page and in the documentation, is the following way:

```javascript

useStore(useShallow((state) => [state.X1,state.X2]));

```

Which one should I use? Or are the two usages serving different purposes?

**Comment by dai-shi** [comment, marked as answer]
Both are introduced in v4 and will stay in v5.
useShallow is newer and is basically recommended.
Basically we want to collect more feedback on them.

**Reply by oosawy** [reply]
I can't make sense of useShallow, although useStore accepts the equality function.

That useShallow usage seems confusing to use a hook in hook arguments, and passing `shallow` as equalityFn should be straightforward.

Why useShallow is recomended? What is the difference?

**Reply by dai-shi** [reply]
#1937 this long thread might help.

In short, some people feel it better. If you want to stick with `shallow`, you can use `createWithEqualityFn`.
Note that `useFoo(useBar(...))` is totally valid usage.

**Reply by alpor** [reply]
>useShallow is newer and is basically recommended

Newer does not means better. Moreover, many todays "newer" things are actually worse than older.
Why it is recommended? You actually sacrifice encapsulation and code reuse. For what?

>this long thread might help

No, it is not.

**Reply by dai-shi** [reply]
> Why it is recommended?

*If both useShallow and shallow work for your case*, useShallow is recommended because it reduces the bundle size. Yes, as it's new we may have unknown issues, that's why it's "basically".

**Comment by hamitkera** [comment]
I created useShallowStore hook to pick any state with just array and TS support. Do you think it's good practise? @dai-shi 
![image](https://github.com/pmndrs/zustand/assets/162107754/c80ef216-8f86-470c-9741-91211e787f38)

**Reply by dai-shi** [reply]
I'm not sure if it's good or recommended, but it's valid.

If I were to do it, I'd simply do this:
```js
const Input = () => {
  const unknown = useStore((state) => state.unknown)
  const name = useStore((state) => state.name)
```

**Reply by hamitkera** [reply]
Sometimes states is growing and I don't want to pick every state line by line so I thought it would be good to have such a hook to pick all at once.

**Reply by dbritto-dev** [reply]
```ts
const [unknown, name] = useStoreWithEqualityFn((state) => [state.unknown, state.name], shallow)
```

**Reply by hamitkera** [reply]
I know this one but just wanted to make it look clean with a hook. Also the purpose of putting unknow there to show that typescript complains if the key is not part of state.

**Comment by oleksandr-danylchenko** [comment]
Do I understand correctly that for the **_vanilla stores_** made with the [`createStore`](https://zustand.docs.pmnd.rs/apis/create-store), using the `useStore` with the `useShallow` is an expected pattern?
```tsx
export function useAnnotatorStore<T>(selector: (state: AnnotatorSlices) => T): T {
	const store = useContext(AnnotatorStoreContext);
	if (!store) {
		throw new Error('Missing `AnnotatorStoreContext.Provider` in the tree');
	}
	return useStore(store, selector);
}
```
```tsx
const { isNarrow } = useAnnotatorStore(useShallow((state) => ({ isNarrow: state.isNarrow })));
```

**Reply by oleksandr-danylchenko** [reply]
I saw in the docs[^1] that the `useStore` hook can be bound to a certain store type using the overrides:
```ts
function useBearStore(): BearState
function useBearStore<T>(selector: (state: BearState) => T): T
function useBearStore<T>(selector?: (state: BearState) => T) {
  return useStore(bearStore, selector!)
}
```
So it would be nice to use such a pattern for the vanilla store distributed through the React Context.

[^1]: https://zustand.docs.pmnd.rs/guides/advanced-typescript#bounded-usestore-hook-for-vanilla-stores

**Reply by dbritto-dev** [reply]
@oleksandr-danylchenko `useShallow` is not required but recommended when you want to return and `object` or `array` but not needed when you return primitive values

**Reply by oleksandr-danylchenko** [reply]
Thanks, sure ✅ 

I was interested specifically in the objects/arrays creation within the selector.

I also use them quite frequently as a single "groped" selector per component:
```ts
const { isNoteEditMode, onNoteCancel, draftHtml } = useAnnotatorStore(
	useShallow((state) => ({
		isNoteEditMode: state.noteMode === NoteMode.EDIT,
		onNoteCancel: state.onNoteCancel,
		draftHtml: state.draftHtml
	}))
);
```
